### ISSUE TITLE
Diff command without ORM

### ISSUE THREAD

**Issue by veewee** [doctrine/migrations#1047]
### Diff command without ORM


|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | no

#### Summary

Currently the diff command is only added when an entity manager is available:

https://github.com/doctrine/migrations/blob/a04b141ee90d630d7b86e7334ebbb77b51a3275a/lib/Doctrine/Migrations/Tools/Console/ConsoleRunner.php#L125-L129

However, the only reason why the diff command could not work is because the `DependencyFactory` does not know how to build the desired schema:

https://github.com/doctrine/migrations/blob/a04b141ee90d630d7b86e7334ebbb77b51a3275a/lib/Doctrine/Migrations/DependencyFactory.php#L238-L243


If you allow a way to insert a custom schema provider in there, you could also make the diff command work for regular DBAL implementations.


How you can implement this in your codebase:

1.  Application logic outide of this package : Add e.g. a table builder like this:

```php
use Doctrine\DBAL\Schema\Table;
use Doctrine\DBAL\Types\Types;

final class SomeTable
{
    public const NAME = 'some_table';

    public static function createTable(): Table
    {
        $table = new Table(self::NAME);

        $table->addOption('charset', 'utf8mb4');
        $table->addOption('collate', 'utf8mb4_unicode_ci');
        $table->addOption('engine', 'InnoDB');

        $table->addColumn('id', Types::GUID, [
            'notnull' => true,
        ]);

        // ....

        $table->setPrimaryKey(['id']);

        return $table;
    }
}
```

(Extra benefit for implementing these table schema's inside your codebase : you can use these table objects as source for e.g. database integration tests, making sure they are always up to date)

2. Inside this package : Create a custom schema provider, which basically looks like this (and could be loaded from the application with e.g. tagged services or a directory loader):

```php
class DependencyFactory
{
// ....

    private function getSchemaProvider() : SchemaProvider
    {
        return $this->getDependency(SchemaProvider::class, function (): SchemaProvider {
            return new class implements SchemaProvider
            {
                public function createSchema(): Schema
                {
                    $schema = new Schema(
                        [
                            SomeTable::createTable(),
                        ]
                    );

                    return $schema;
                }
            };
        });
    }

// ....
}
```

The changes above result in:

```sh
./bin/console migrations:diff                                                                                                                                                                                                      

// ....

 Generated new migration class to "/.../migrations/Version20200817092354.php"

// ....
```


```php
final class Version20200817092354 extends AbstractMigration
{
    public function getDescription() : string
    {
        return '';
    }

    public function up(Schema $schema) : void
    {
        // this up() migration is auto-generated, please modify it to your needs
        $this->addSql('CREATE TABLE some_table (id CHAR(36) NOT NULL COMMENT \'(DC2Type:guid)\', PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
    }

    public function down(Schema $schema) : void
    {
        // this down() migration is auto-generated, please modify it to your needs
        $this->addSql('DROP TABLE some_table');
    }
}
```

So it seems quite trivial to add a way to provide your own schema in order to generate a diff.
What do you think about this feature and what's the best approach to make the schema configurable inside the `DependencyFactory`?

**Comment by goetas** [comment]
Hmm, that makes sense. The diff command should be available based on the availability of the schema provider, not on the entity manager

**Comment by veewee** [comment]
Fixed in #1050
